modules/tensor_mechanics/src/userobjects/CrystalPlasticityStateVarRateComponentVoce.C

```diff
@@ -0,0 +1,288 @@
+               "parameters does not match the number of slip system groups");
+
+  // initialize useful variables;
+  initSlipSystem_PlaneID(_slipSystem_PlaneID);
+  initSlipSystem_GroupID(_slipSystem_GroupID);
+}
+
+bool
+CrystalPlasticityStateVarRateComponentVoce::calcStateVariableEvolutionRateComponent(
+    unsigned int qp, std::vector<Real> & val) const
+{
+  val.assign(_variable_size, 0.0);
+
+  unsigned int group_i;
+  Real h0;
+  Real tau_0;
+  Real tau_sat;
+  Real hardening_exponenet;
+  Real delta_tau;
+
+  DenseVector<Real> hb(_variable_size);
+
+  for (unsigned int i = 0; i < _variable_size; ++i)
+  {
+    group_i = _slipSystem_GroupID[i];
+    h0 = _h0_group_values[group_i];
+    tau_0 = _tau0_group_values[group_i];
+    tau_sat = _tauSat_group_values[group_i];
+    hardening_exponenet = _hardeningExponent_group_values[group_i];
+
+    delta_tau = tau_sat - tau_0;
+
+    hb(i) = h0 *
+            std::pow(std::abs(1.0 - (_mat_prop_state_var[qp][i] - tau_0) / delta_tau),
+                     hardening_exponenet) *
+            std::copysign(1.0, 1.0 - (_mat_prop_state_var[qp][i] - tau_0) / delta_tau);
+  }
+
+  for (unsigned int i = 0; i < _variable_size; ++i)
+    for (unsigned int j = 0; j < _variable_size; ++j)
+    {
+
+      Real q_ab = getHardeningCoefficient(i, j);
+
+      val[i] += std::abs(_mat_prop_slip_rate[qp][j]) * q_ab * hb(j);
+    }
+
+  return true;
+}
+
+MooseEnum
+CrystalPlasticityStateVarRateComponentVoce::crystalLatticeTypeOptions()
+{
+  return MooseEnum("FCC BCC", "FCC");
+}
+
+void
+CrystalPlasticityStateVarRateComponentVoce::initSlipSystem_PlaneID(
+    std::vector<unsigned int> & _slipSystem_PlaneID) const
+{
+  // this routine is generate a vector containing the association between
+  // slip system number and slip plane
+  _slipSystem_PlaneID.assign(_variable_size, 0);
+
+  for (unsigned int slipSystemIndex = 0; slipSystemIndex < _variable_size; ++slipSystemIndex)
+  {
+
+    switch (_crystal_lattice_type)
+    {
+      case 0: // FCC
+        if (slipSystemIndex < 12)
+          _slipSystem_PlaneID[slipSystemIndex] = slipSystemIndex / 3;
+        else
+          mooseError("FCC with more than 12 slip planes is not implemented ");
+
+        break;
+
+      case 1: // BCC
+        if (slipSystemIndex < 12)
+          _slipSystem_PlaneID[slipSystemIndex] = slipSystemIndex / 2;
+
+        else if (slipSystemIndex >= 12 && slipSystemIndex < 48)
+          _slipSystem_PlaneID[slipSystemIndex] = (slipSystemIndex - 6);
+
+        else
+          mooseError("BCC with more than 48 slip systems is not implemented ");
+
+        break;
+
+      default:
+        mooseError("VoceHardeningError: Pass valid crustal_structure_type ");
+    }
+  }
+}
+
+void
+CrystalPlasticityStateVarRateComponentVoce::initSlipSystem_GroupID(
+    std::vector<unsigned int> & _slipSystem_GroupID) const
+// this routine is generate a vector containing the association between
+// slip system number and provided group edges
```

Comment:
this comment seems out of place. I suggest putting it into the header file where the member is declared. That way it gets picked up by the doxugen documentation system (make sure you use `///` triple-slash comment)